### PR TITLE
Add CyclicBarrier

### LINKED ISSUE #5801: Add CyclicBarrier
>A synchronization aid that allows a set of fibers to all wait for each other to reach a common barrier point. CyclicBarriers are useful in programs involving a fixed sized party of fibers that must occasionally wait for each other. The barrier is called cyclic because it can be re-used after the waiting fibers are released.

```scala
package zio.concurrent

trait CyclicBarrier:
  def parties: Int
  def waiting: UIO[Int]
  def await: IO[Unit, Unit]
  def reset: UIO[Unit]
  def isBroken: UIO[Boolean]

object CyclicBarrier:
  def make(n: Int): UIO[CyclicBarrier]
```

### PR DESCRIPTION
**Pull request by regiskuckaertz**
Closes #5801 

@adamgfraser @mijicd The action is quite restrictive, what are your thoughts on this?

### CODE REVIEW

File: core/shared/src/main/scala/zio/concurrent/CyclicBarrier.scala [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+  /** The number of parties required to trip this barrier. */
+  def parties: Int = _parties
+
+  /** The number of parties currently waiting at the barrier. */
+  val waiting: UIO[Int] = _waiting.get
+
+  /** Waits until all parties have invoked await on this barrier. Fails if the barrier is broken. */
+  val await: IO[Unit, Int] =
+    _broken.get.flatMap(if (_) IO.fail(()) else UIO.unit) *>
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/zio/concurrent/CyclicBarrier.scala:39]: Should the whole action be uninterruptible? (except for `action` probably)

File: core/shared/src/main/scala/zio/concurrent/CyclicBarrier.scala [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+ * A synchronization aid that allows a set of fibers to all wait for each other to reach a common barrier point.
+ *
+ * CyclicBarriers are useful in programs involving a fixed sized party of fibers that must occasionally wait for each
+ * other. The barrier is called cyclic because it can be re-used after the waiting fibers are released.
+ *
+ * A CyclicBarrier supports an optional action command that is run once per barrier point, after the last fiber in the
+ * party arrives, but before any fibers are released. This barrier action is useful for updating shared-state before
+ * any of the parties continue.
+ */
+final class CyclicBarrier(
```
Comment (mijicd) [inline core/shared/src/main/scala/zio/concurrent/CyclicBarrier.scala:15]: Does it makes sense to make the constructor private?

File: core/shared/src/main/scala/zio/concurrent/CyclicBarrier.scala [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  private val _broken: Ref[Boolean]
+) {
+  private val break: UIO[Unit] =
+    _broken.set(true) *> fail
+
+  private val fail: UIO[Unit] =
+    _lock.get.flatMap(_.fail(()).unit)
+
+  private val succeed: UIO[Unit] =
+    _lock.get.flatMap(_.succeed(()).unit)
```
Comment (mijicd) [inline core/shared/src/main/scala/zio/concurrent/CyclicBarrier.scala:29]: I'd define these as `UIO[Any]`.

File: core/shared/src/main/scala/zio/concurrent/CyclicBarrier.scala (hunk 2 above)
Comment (regiskuckaertz) [reply core/shared/src/main/scala/zio/concurrent/CyclicBarrier.scala:15]: Yes, good catch

File: core/shared/src/main/scala/zio/concurrent/CyclicBarrier.scala (hunk 1 above)
Comment (regiskuckaertz) [reply core/shared/src/main/scala/zio/concurrent/CyclicBarrier.scala]: Done

File: core/shared/src/main/scala/zio/concurrent/CyclicBarrier.scala (hunk 2 above)
Comment (regiskuckaertz) [reply core/shared/src/main/scala/zio/concurrent/CyclicBarrier.scala]: Done

File: core/shared/src/main/scala/zio/concurrent/CyclicBarrier.scala (hunk 3 above)
Comment (regiskuckaertz) [reply core/shared/src/main/scala/zio/concurrent/CyclicBarrier.scala]: Done

**Review by mijicd — APPROVED** [review]

**Review by mijicd — DISMISSED** [review]

**Review by mijicd — APPROVED** [review]

**Pull request merged**